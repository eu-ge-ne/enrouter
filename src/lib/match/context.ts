import { createContext, useContext } from "react";

import type { Match } from "./match.js";

const MatchesContext = createContext<Match[]>([]);
const MatchIndexContext = createContext<number>(-1);

export const MatchesProvider = MatchesContext.Provider;
export const MatchIndexProvider = MatchIndexContext.Provider;

export interface MatchContext {
  matches: Match[];
  matchIndex: number;
  match: Match | undefined;
  firstMatch: Match | undefined;
  nextMatch: Match | undefined;
  lastMatch: Match | undefined;
  voidMatch: Match | undefined;
  isExactMatch: boolean;
}

export function useMatchContext(): MatchContext {
  const matches = useContext(MatchesContext);
  const matchIndex = useContext(MatchIndexContext);

  const match = matches[matchIndex];
  const firstMatch = matches[0];
  const nextMatch = matches[matchIndex + 1];
  const lastMatch = matches.at(-1);
  const voidMatch = matches.findLast((x) => x.route?.elements._void);

  const isExactMatch = Boolean(lastMatch?.route);

  return {
    matches,
    matchIndex,
    match,
    firstMatch,
    nextMatch,
    lastMatch,
    voidMatch,
    isExactMatch,
  };
}

export function useMatch(path: string): Match | undefined {
  return useContext(MatchesContext).find((x) => x.route?.path === path);
}
