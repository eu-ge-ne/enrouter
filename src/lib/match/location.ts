import type { Route } from "#lib/route/mod.js";
import type { Match } from "./mod.js";
import { logger } from "#lib/debug.js";
import { getRouteTree } from "#lib/route/tree.js";
import { loadRoutes } from "#lib/route/load.js";

const log = logger("matchLocation");

export async function matchLocation(
  location: string,
): Promise<Match | undefined> {
  const matches: Match[] = [];

  recur([getRouteTree()], location, matches);

  await loadRoutes(matches.map((x) => x.route));

  link(matches);

  log("matched: %o", matches);

  return matches[0];
}

function recur(routes: Route[], location: string, matches: Match[]): void {
  const match = matchOneOf(routes, location);
  if (!match) {
    return;
  }

  log(`path: "%s", location: "%s"`, match.route.path, match.location);

  matches.push(match);

  if (match.route.tree) {
    recur(match.route.tree, location, matches);
  }
}

function matchOneOf(routes: Route[], location: string): Match | undefined {
  let matched = routes
    .map((route) => {
      const execs = route.test.pattern.exec(location);
      if (execs) {
        return { route, execs };
      }
    })
    .filter((x) => x !== undefined);

  if (matched.length === 0) {
    return;
  }

  // TODO: improve

  if (matched.length > 1) {
    matched = matched.filter((x) => x.route.test.keys.length === 0);
  }

  if (matched.length !== 1) {
    throw new Error("Found >1 match");
  }

  const { route, execs } = matched[0]!;
  const loc = execs[0] || "/";

  return {
    isVoid: false,
    route,
    isExact: loc === location,
    location: loc,
    params: Object.fromEntries(
      route.test.keys.map((key, i) => [key, execs![i + 1]!] as const),
    ),
  };
}

function link(matches: Match[]): void {
  const lastVoidIndex = matches.findLastIndex((x) => x.route.elements._void);

  const first = matches[0];
  const last = matches.at(-1);

  matches.forEach((x, i) => {
    x.isVoid = i === lastVoidIndex;

    x.first = first;
    x.prev = matches[i - 1];
    x.next = matches[i + 1];
    x.last = last;
  });
}
