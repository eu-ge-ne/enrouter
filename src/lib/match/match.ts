import type { Route } from "#lib/route/mod.js";
import { getRouteTree } from "#lib/route/tree.js";
import { loadRoutes } from "#lib/route/load.js";

export interface Match {
  route?: Route;
  location: string;
  params: Record<string, string>;
}

export async function matchLocation(location: string): Promise<Match[]> {
  const matches: Match[] = [];

  let routes: Route[] | undefined = [getRouteTree()];

  while (routes) {
    const match = matchRoute(routes, location);

    routes = match?.route?.tree;

    if (match) {
      matches.push(match);
    }
  }

  await loadRoutes(matches.map((x) => x.route!));

  if (matches.at(-1)?.location !== location) {
    matches.push({ location, params: {} });
  }

  return matches;
}

function matchRoute(routes: Route[], location: string): Match | undefined {
  let matched = routes
    .map((route) => {
      const execs = route.test.pattern.exec(location);
      if (execs) {
        return { route, execs };
      }
    })
    .filter((x) => x !== undefined);

  if (matched.length === 0) {
    return;
  }

  // TODO: improve

  if (matched.length > 1) {
    matched = matched.filter((x) => x.route.test.keys.length === 0);
  }

  if (matched.length !== 1) {
    throw new Error("Found >1 match");
  }

  const { route, execs } = matched[0]!;

  return {
    route,
    location: execs[0] || "/",
    params: Object.fromEntries(
      route.test.keys.map((key, i) => [key, execs![i + 1]!] as const),
    ),
  };
}
