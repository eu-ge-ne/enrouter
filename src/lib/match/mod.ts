import type { Route } from "#lib/route/mod.js";

/**
 * Represents matched segment of location.
 */
export interface Match {
  isVoid: boolean;
  route: Route;

  isExact: boolean;
  location: string;
  params: Record<string, string>;

  first?: Match;
  prev?: Match;
  next?: Match;
  last?: Match;
}
