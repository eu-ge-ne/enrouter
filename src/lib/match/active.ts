import { useMatches } from "./context.js";

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({
  path,
  loose,
  value: [yes, no],
}: UseActiveParams<T>): T {
  const matches = useMatches();

  const match = matches.find((x) => x.route?.path === path);

  if (!match?.route) {
    return no;
  }

  if (loose) {
    return yes;
  }

  if (match !== matches.at(-1)) {
    return no;
  }

  return match?.route ? yes : no;
}
