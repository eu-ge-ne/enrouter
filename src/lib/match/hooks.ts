import type { Match } from "./mod.js";
import { useMatch } from "./context.js";

export function useRoot(): Match | undefined {
  let match = useMatch()?.last;

  while (match) {
    if (match.route.elements._layout?.Root) {
      return match;
    }
    match = match.prev;
  }
}

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({ path, loose, value }: UseActiveParams<T>): T {
  const match = useMatch(path);

  const isActive = loose ? match !== undefined : Boolean(match?.isExact);

  return isActive ? value[0] : value[1];
}
