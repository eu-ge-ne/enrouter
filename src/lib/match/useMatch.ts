import { useContext } from "react";

import type { Match } from "./match.js";
import { MatchesContext, MatchIndexContext } from "./context.js";

export interface UseMatchResult {
  current: Match | undefined;
  first: Match | undefined;
  next: Match | undefined;
  last: Match | undefined;
  fallback: Match | undefined;
  isExact: boolean;
}

export function useMatch(): UseMatchResult {
  const matches = useContext(MatchesContext);
  const matchIndex = useContext(MatchIndexContext);

  const current = matches[matchIndex];
  const first = matches[0];
  const next = matches[matchIndex + 1];
  const last = matches.at(-1);
  const fallback = matches.findLast((x) => x.route?.elements._fallback);

  const isExact = Boolean(last?.route);

  return {
    current,
    first,
    next,
    last,
    fallback,
    isExact,
  };
}
