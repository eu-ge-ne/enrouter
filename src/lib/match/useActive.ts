import { useMatch } from "./useMatch.js";
import { useMatchFor } from "./useMatchFor.js";

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({
  path,
  loose,
  value: [yes, no],
}: UseActiveParams<T>): T {
  const match = useMatchFor(path);
  const { last } = useMatch();

  if (!match?.route) {
    return no;
  }

  if (loose) {
    return yes;
  }

  if (match !== last) {
    return no;
  }

  return match?.route ? yes : no;
}
