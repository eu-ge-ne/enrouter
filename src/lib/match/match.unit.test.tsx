import { describe, test, expect, vi } from "vitest";
import * as regexparam from "regexparam";

import { getRouteTree } from "#lib/route/tree.js";
import { matchLocation } from "./match.js";

vi.mock(import("#lib/route/tree.js"), () => ({
  getRouteTree: vi.fn(),
}));

describe("match", () => {
  describe("matchLocation", () => {
    test("0 matches", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/abc",
        test: regexparam.parse("/abc", true),
        modules: [],
        loaded: true,
        elements: {},
      });

      expect(await matchLocation("/x")).toMatchSnapshot();
    });

    test("1 match", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: true,
        elements: {},
      });

      expect(await matchLocation("/")).toMatchSnapshot();
    });

    test("2 matches", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: true,
        elements: {},
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {},
          },
        ],
      });

      expect(await matchLocation("/abc")).toMatchSnapshot();
    });

    test("1 match with params", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/[:id]",
        test: regexparam.parse("/:id", true),
        modules: [],
        loaded: true,
        elements: {},
      });

      expect(await matchLocation("/100")).toMatchSnapshot();
    });

    test("1 match and 1 match with params", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: true,
        elements: {},
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {},
          },
          {
            path: "/[:id]",
            test: regexparam.parse("/:id", true),
            modules: [],
            loaded: true,
            elements: {},
          },
        ],
      });

      expect(await matchLocation("/abc")).toMatchSnapshot();
    });

    test("3 matches and 1 not found", async () => {
      vi.mocked(getRouteTree).mockReturnValueOnce({
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: true,
        elements: {},
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {},
            tree: [
              {
                path: "/abc/xyz",
                test: regexparam.parse("/abc/xyz", true),
                modules: [],
                loaded: true,
                elements: {},
              },
            ],
          },
        ],
      });

      expect(await matchLocation("/abc/xyz/100")).toMatchSnapshot();
    });
  });
});
