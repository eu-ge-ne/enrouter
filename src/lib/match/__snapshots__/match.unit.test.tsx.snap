// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`match > matchLocation > 0 matches 1`] = `
[
  {
    "location": "/x",
    "params": {},
  },
]
`;

exports[`match > matchLocation > 1 match 1`] = `
[
  {
    "location": "/",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > matchLocation > 1 match with params 1`] = `
[
  {
    "location": "/100",
    "params": {
      "id": "100",
    },
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/[:id]",
      "test": {
        "keys": [
          "id",
        ],
        "pattern": /\\^\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > matchLocation > 2 matches 1`] = `
[
  {
    "location": "/",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/abc",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
      ],
    },
  },
  {
    "location": "/abc",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/abc",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > matchLocation > 2 matches with params 1`] = `
[
  {
    "location": "/",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/abc",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
        {
          "elements": {},
          "loaded": false,
          "modules": [],
          "path": "/[:id]",
          "test": {
            "keys": [
              "id",
            ],
            "pattern": /\\^\\\\/\\(\\[\\^/\\]\\+\\?\\)\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
      ],
    },
  },
  {
    "location": "/abc",
    "params": {},
    "route": {
      "elements": {},
      "loaded": false,
      "modules": [],
      "path": "/abc",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
]
`;

exports[`match > matchLocation > 3 matches and 1 not found 1`] = `
[
  {
    "location": "/",
    "params": {},
    "route": {
      "elements": {},
      "loaded": true,
      "modules": [],
      "path": "/",
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": true,
          "modules": [],
          "path": "/abc",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
          },
          "tree": [
            {
              "elements": {},
              "loaded": true,
              "modules": [],
              "path": "/abc/xyz",
              "test": {
                "keys": [],
                "pattern": /\\^\\\\/abc\\\\/xyz\\(\\?=\\$\\|\\\\/\\)/i,
              },
            },
          ],
        },
      ],
    },
  },
  {
    "location": "/abc",
    "params": {},
    "route": {
      "elements": {},
      "loaded": true,
      "modules": [],
      "path": "/abc",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\(\\?=\\$\\|\\\\/\\)/i,
      },
      "tree": [
        {
          "elements": {},
          "loaded": true,
          "modules": [],
          "path": "/abc/xyz",
          "test": {
            "keys": [],
            "pattern": /\\^\\\\/abc\\\\/xyz\\(\\?=\\$\\|\\\\/\\)/i,
          },
        },
      ],
    },
  },
  {
    "location": "/abc/xyz",
    "params": {},
    "route": {
      "elements": {},
      "loaded": true,
      "modules": [],
      "path": "/abc/xyz",
      "test": {
        "keys": [],
        "pattern": /\\^\\\\/abc\\\\/xyz\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
  },
  {
    "location": "/abc/xyz/100",
    "params": {},
  },
]
`;
